### PR TITLE
CommonITIL: fix restore input on failed submission

### PR DESCRIPTION
**Pull request by AdrienClairembault**
When a ticket submission fails (e.g. missing mandatory fields), the user submitted data is lost (ticket title, description, ...).

image

I've found that this is due to a code change from GLPI 9.5 to GLPI 10.0.
The ticket form may be reloaded from different actions:
- Failed submission (missing mandatory fields)
- Important data change (entity, category, requester, ...)

The backend handle the saved input differently for these two cases and thus must be able to differentiate them.
To do this, it rely on `$options`:
- Failed submission -> `$option` will be empty
- Data change -> `$options` will contain the form data written by the user

With this in mind, the backend only need to check if  `$option` is empty to do the right decision.

This was broken in GLPI 10 because `$options` is no longer empty in case of a failed submission:
```
[2022-11-14 14:19:21] glpiphplog.DEBUG: CommonGLPI::showTabsContent() in /home/aclai/localhost/glpi/10.0-bugfixes/src/CommonGLPI.php line 846
Array
  (
      [id] => 0
  )
``` 

To fix this, I've emptied `$options['id']` in this specific case.

image

Obviously  relying on an empty `$options` is not reliable.
A better fix would be to rework the backend to not longer "guess" which action was taken but rather get a specific instruction from the front.

I've seen that a first attempt was done here but was later reverted: #12929 
Maybe something to add to GLPI 10.1 roadmap if it's not already done ?

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !25540

### CODE REVIEW

**Review by trasher — APPROVED** [review]

**Review by stonebuzz — APPROVED** [review]
Seems good for me ! Tested 

- on submit failed (mandatory field not filled on submit)
- on reload form (with dropdown category)
- with fields plugins
- large content

**Review by cedric-anne — COMMENTED** [review]
The `saveInput()` call is made using `$cleaned_options` that has already its `id` key dropped. It may be easier to understand to just do the check related to options count using `$cleaned_options`.
```diff
diff --git a/src/CommonGLPI.php b/src/CommonGLPI.php
index a8133d0b2a..bc5ed1134a 100644
--- a/src/CommonGLPI.php
+++ b/src/CommonGLPI.php
@@ -856,18 +856,16 @@ class CommonGLPI implements CommonGLPIInterface
 
         $target         = $_SERVER['PHP_SELF'];
         $extraparamhtml = "";
-        $withtemplate   = "";
-        if (is_array($options) && count($options)) {
-            if (isset($options['withtemplate'])) {
-                $withtemplate = $options['withtemplate'];
-            }
-            $cleaned_options = $options;
-            if (isset($cleaned_options['id'])) {
-                unset($cleaned_options['id']);
-            }
-            if (isset($cleaned_options['stock_image'])) {
-                unset($cleaned_options['stock_image']);
-            }
+        $withtemplate   = $options['withtemplate'] ?? "";
+
+        $cleaned_options = $options;
+        if (isset($cleaned_options['id'])) {
+            unset($cleaned_options['id']);
+        }
+        if (isset($cleaned_options['stock_image'])) {
+            unset($cleaned_options['stock_image']);
+        }
+        if (count($cleaned_options) > 0) {
             if ($this instanceof CommonITILObject && $this->isNewItem()) {
                 $this->input = $cleaned_options;
                 $this->saveInput()
```

File: src/CommonGLPI.php [hunk 1]
```diff
@@ -854,20 +854,29 @@ public function showTabsContent($options = [])
             }
         }
 
+        $cleaned_options = $options;
+        if (isset($cleaned_options['id'])) {
+            unset($cleaned_options['id']);
+        }
+        if (isset($cleaned_options['stock_image'])) {
+            unset($cleaned_options['stock_image']);
+        }
```
Comment (cedric-anne) [inline src/CommonGLPI.php:863]: Unset can take multiple arguments and will not trigger any warning if array keys are not set.
```suggestion
        unset($cleaned_options['id'], $cleaned_options['stock_image']);
```

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]
Seems ok on my side.
As already said, we strinly need to clean all this part.
It's mess

**Review by trasher — APPROVED** [review]

**Pull request merged**